File: androidApp/src/main/java/com/sbellanger/weather/android/presentation/WeatherScreen.kt
package com.sbellanger.weather.android.presentation

import android.content.res.Configuration.UI_MODE_NIGHT_YES
import androidx.compose.animation.ExperimentalAnimationApi
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.wrapContentSize
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.sp
import com.sbellanger.weather.ViewState
import com.sbellanger.weather.android.R

@ExperimentalAnimationApi
@Composable
fun WeatherScreen(viewState: ViewState) {
    Box {
        Image(
            painter = painterResource(id = R.drawable.background_night),
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier
                .fillMaxSize()
        )
        when (viewState) {
            is ViewState.Error -> ErrorScreen(viewState)
            is ViewState.HasResult -> LoadedScreen(viewState)
            ViewState.Loading -> LoadingScreen()
        }
    }
}

@Composable
fun LoadedScreen(viewState: ViewState.HasResult) {
    WeatherDescriptionView(
        temperature = viewState.weather.temperature,
        city = viewState.weather.city,
        date = viewState.weather.date,
        description = viewState.weather.description,
        humidity = viewState.weather.humidity,
        pressure = viewState.weather.pressure,
        wind = viewState.weather.wind
    )
}

@Composable
fun ErrorScreen(viewState: ViewState.Error) {
    Text(
        text = viewState.errorMessage,
        modifier = Modifier
            .fillMaxWidth()
            .wrapContentSize(Alignment.Center),
        color = Color.White,
        fontWeight = FontWeight.Bold,
        fontSize = 50.sp
    )
}

@ExperimentalAnimationApi
@Composable
fun LoadingScreen() {
    LoaderView(true)
}

@ExperimentalAnimationApi
@Preview(showBackground = true)
@Composable
fun DefaultPreview() {
    WeatherScreen(ViewState.Loading)
}

@ExperimentalAnimationApi
@Preview(
    showBackground = true,
    uiMode = UI_MODE_NIGHT_YES
)
@Composable
fun DarkModePreview() {
    WeatherScreen(ViewState.Loading)
}

File: iosApp/iosApp/Feature/WeatherScreen.swift
//
//  WeatherScreen.swift
//  iosApp
//
//  Created by Stephen BELLANGER on 09/03/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import SwiftUI
import sha

struct WeatherScreen : View {
    let viewModel: WeatherViewModel
    
    var body: some View {
        GeometryReader { geometry in
            ZStack {
                Image("background_night")
                    .resizable()
                    .scaledToFill()
                    .edgesIgnoringSafeArea(.all)
                if(viewModel.viewState is ViewState.HasResult){
                    let successState = viewModel.viewState as! ViewState.HasResult
                    WeatherDescriptionView(
                        temperature: successState.weather.temperature,
                        city: successState.weather.city,
                        date: successState.weather.date,
                        description: successState.weather.description_,
                        humidity: successState.weather.humidity,
                        pressure: successState.weather.pressure,
                        wind: successState.weather.wind
                    )
                } else if(viewModel.viewState is ViewState.Error){
                    let errorState = viewModel.viewState as! ViewState.Error
                    ErrorView(message: errorState.errorMessage)

                } else if (viewModel.viewState is ViewState.Loading){
                    LoaderView()
                }
            }
        }
    }
}
